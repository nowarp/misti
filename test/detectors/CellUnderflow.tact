fun refs_num_bad(c: Cell) {
    let s1 = beginCell().storeRef(c).endCell().asSlice();
    let ref1 = s1.loadRef(); // OK
    let ref2 = s1.loadRef(); // Bad: Cell Underflow

    // Bad: Cell Underflow
    beginCell()
        .endCell()
        .asSlice()
        .loadRef();
}
