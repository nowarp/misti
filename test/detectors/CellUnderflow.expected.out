[CRITICAL] CellUnderflow: Reference count might go below 0
test/detectors/CellUnderflow.tact:2:9:
  1 | fun refs_underflow(c: Cell) {
> 2 |     let s1 = beginCell()
              ^
  3 |                .storeRef(c)
The possible number of references stored (1) is less than loaded (2)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellUnderflow

[CRITICAL] CellUnderflow: Reference count might go below 0
test/detectors/CellUnderflow.tact:7:9:
  6 |     let ref1 = s1.loadRef(); // OK
> 7 |     let ref2 = s1.loadRef(); // Bad: Cell Underflow
              ^
  8 | 
The possible number of references stored (1) is less than loaded (2)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellUnderflow

[CRITICAL] CellUnderflow: Reference count might go below 0
test/detectors/CellUnderflow.tact:9:5:
   8 | 
>  9 |     beginCell() // Bad: Cell Underflow
           ^
  10 |         .endCell()
The possible number of references stored (0) is less than loaded (1)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellUnderflow

[CRITICAL] CellUnderflow: Too many references stored in cell
test/detectors/CellUnderflow.tact:22:9:
  21 |                .storeRef(c); // OK: Builder storing 4 references
> 22 |     let c1 = beginCell() // Bad: We got 5 references
               ^
  23 |                .storeBuilder(b1)
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellUnderflow

[CRITICAL] CellUnderflow: Too many references stored in cell
test/detectors/CellUnderflow.tact:27:9:
  26 | 
> 27 |     let c2 = // Bad: 4+ refs
               ^
  28 |     beginCell()
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellUnderflow

[CRITICAL] CellUnderflow: Too many references stored in cell
test/detectors/CellUnderflow.tact:41:5:
  40 |     b2 = b2.storeRef(c);
> 41 |     b2 = b2.storeRef(c); // Bad: 4+ refs
           ^
  42 |     let c3 = b2.endCell(); // Bad: 4+ refs
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellUnderflow

[CRITICAL] CellUnderflow: Too many references stored in cell
test/detectors/CellUnderflow.tact:42:9:
  41 |     b2 = b2.storeRef(c); // Bad: 4+ refs
> 42 |     let c3 = b2.endCell(); // Bad: 4+ refs
               ^
  43 | }
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellUnderflow